libs/ballot-interpreter-nh-next/src/timing_marks.rs

```diff
@@ -0,0 +1,1038 @@
+    let next_point_vector = segment.with_length(expected_distance).vector();
+    let maximum_error = expected_distance / 2.0;
+    while inferred_timing_marks.len() < expected_count as usize {
+        // find the closest existing timing mark
+        let closest_rect = timing_marks
+            .iter()
+            .min_by(|a, b| {
+                let a_distance = Segment::new((a).center(), current_timing_mark_center).length();
+                let b_distance = Segment::new((b).center(), current_timing_mark_center).length();
+                a_distance
+                    .partial_cmp(&b_distance)
+                    .unwrap_or(std::cmp::Ordering::Equal)
+            })
+            .map_or_else(
+                || unreachable!("there will always be a closest timing mark"),
+                |rect| rect,
+            );
+
+        // if the closest timing mark is close enough, use it
+        if Segment::new((closest_rect).center(), current_timing_mark_center).length()
+            <= maximum_error
+        {
+            inferred_timing_marks.push(*closest_rect);
+            current_timing_mark_center = (closest_rect).center() + next_point_vector;
+        } else {
+            // otherwise, we need to fill in a point
+            inferred_timing_marks.push(Rect::new(
+                (current_timing_mark_center.x - geometry.timing_mark_size.width / 2.0).round()
+                    as i32,
+                (current_timing_mark_center.y - geometry.timing_mark_size.height / 2.0).round()
+                    as i32,
+                geometry.timing_mark_size.width.round() as u32,
+                geometry.timing_mark_size.height.round() as u32,
+            ));
+            current_timing_mark_center += next_point_vector;
+        }
+    }
+    inferred_timing_marks
+}
+
+/// Determines whether a rect could be a timing mark based on its size.
+pub fn rect_could_be_timing_mark(geometry: &Geometry, rect: &Rect) -> bool {
+    let min_timing_mark_width = (geometry.timing_mark_size.width * 1.0 / 4.0).floor() as u32;
+    let max_timing_mark_width = (geometry.timing_mark_size.width * 3.0 / 2.0).ceil() as u32;
+    let min_timing_mark_height = (geometry.timing_mark_size.height * 2.0 / 3.0).floor() as u32;
+    let max_timing_mark_height = (geometry.timing_mark_size.height * 3.0 / 2.0).ceil() as u32;
+    rect.width() >= min_timing_mark_width
+        && rect.width() <= max_timing_mark_width
+        && rect.height() >= min_timing_mark_height
+        && rect.height() <= max_timing_mark_height
+}
+
+/// Gets all the distances between adjacent rects in a list of rects.
+pub fn distances_between_rects(rects: &[Rect]) -> Vec<f32> {
+    let mut distances = rects
+        .windows(2)
+        .map(|w| Segment::new((&w[1]).center(), (&w[0]).center()).length())
+        .collect::<Vec<f32>>();
+    distances.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
+    distances
+}
+
+#[derive(Clone, Serialize)]
```

Comment:
Maybe all the scoring code goes into its own module?

Reply:
Done.